sdk/src/main/java/io/opentelemetry/sdk/common/ConfigBuilder.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.common;
+
+import java.util.Map;
+
+public abstract class ConfigBuilder {
+
+  protected ConfigBuilder() {}
+
+  public abstract ConfigBuilder fromConfigMap(Map<String, String> configMap);
+
+  public abstract ConfigBuilder fromEnv();
+
+  protected boolean getProperty(String name, Map<String, String> map, boolean defaultValue) {
+    try {
+      return Boolean.parseBoolean(System.getProperty(name, map.get(name)));
+    } catch (NumberFormatException ex) {
+      return defaultValue;
+    }
+  }
+
+  protected int getProperty(String name, Map<String, String> map, int defaultValue) {
+    try {
+      return Integer.parseInt(System.getProperty(name, map.get(name)));
+    } catch (NumberFormatException ex) {
+      return defaultValue;
+    }
+  }
+
+  protected long getProperty(String name, Map<String, String> map, long defaultValue) {
+    try {
+      return Long.parseLong(System.getProperty(name, map.get(name)));
+    } catch (NumberFormatException ex) {
+      return defaultValue;
+    }
+  }
```

Comment:
I would not hardcode that we always look in System properties then env then default. I would allow the following workflow:
* Start from default;
* Apply changes from env;
* Apply changes from System properties;

```
BatchSpanProcessor.getBuilder().overwriteFromEnv().overwriteFromSystemProperties().build()
```

The default should be the proposed solution but would like if the users can also change this order if needed.

Thoughts on this?

Reply:
I'd even prefer completely separating thes steps of collection and application of configuration. The builder just has `applyFromMap` and the collection happens separately.

I.e., I propose something like:

```java
final Config cfg = Config.newBuilder().readEnv().readSystemProps().readCommandLine(args).set(MY_KEY, MY_OVERRIDE_VALUE).build();
final BatchSpanProcessor processor = BatchSpanProcessor.newBuilder().applyConfig(cfg).build();
```

Reply:
If we allowed `Map<String, Object>` (or stored that in the hypothetical Configuration class) we could even support different representations of complex objects like resources from different sources, e.g. JSON object in JSON, URL/query encoded in envvar, ....

Reply:
I liked the proposed API to build the config. "nit: applyConfig -> setConfig" probably better name for a builder.

Also I expect that "readX" methods only apply changes to arguments defined by "X" correct? Does not overwrite all the arguments with default if not defined in the config.

We can also offer a helper `Config.loadFromDefaultSources()` (we can think of a better name) which essentially allows users to avoid the 5 lines you wrote to build the config.